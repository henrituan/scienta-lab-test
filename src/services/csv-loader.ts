import fs from 'fs';
import path from 'path';
import { parse } from 'csv-parse';

export const loadCSV = <T>(filename: string): Promise<T[]> => {
  return new Promise((resolve, reject) => {
    const filePath = path.join(process.cwd(), 'public/data', filename);
    const fileContent = fs.readFileSync(filePath, { encoding: 'utf-8' });

    parse(
      fileContent,
      {
        columns: true,
        skip_empty_lines: true,
      },
      (err, records) => {
        if (err) {
          reject(err);
        } else {
          resolve(records);
        }
      },
    );
  });
};
