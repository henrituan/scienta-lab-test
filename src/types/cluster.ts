export type ClusterRawData = {
  cluster_id: string;
  '%female': string;
  avg_age: string;
  symptoms: string;
};

export type Cluster = {
  clusterId: number;
  avgAge: number;
  femalePercent: number;
  malePercent: number;
  symptoms: string[];
};
