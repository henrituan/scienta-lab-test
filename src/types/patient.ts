export type PatientRawData = {
  patient_id: string;
  cluster_id: string;
  coordinates: string;
};

export type Patient = {
  patientId: number;
  clusterId: number;
  coordinates: {
    x: number;
    y: number;
  };
};
