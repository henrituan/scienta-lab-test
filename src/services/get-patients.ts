import { Patient, PatientRawData } from '@/types/patient';

import { loadCSV } from './csv-loader';

export const getPatients = async (): Promise<Patient[]> => {
  const records = await loadCSV<PatientRawData>('patients.csv');

  return records.map((record) => {
    const coordinates = record['coordinates']
      .replace('(', '')
      .replace(')', '')
      .split(',')
      .map(Number);

    return {
      patientId: parseInt(record['patient_id']),
      clusterId: parseInt(record['cluster_id']),
      coordinates: {
        x: coordinates[0],
        y: coordinates[1],
      },
    };
  });
};
