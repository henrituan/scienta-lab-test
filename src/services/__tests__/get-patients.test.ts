import { vi, describe, it, expect } from 'vitest';

import { Patient, PatientRawData } from '@/types/patient';

import { getPatients } from '../get-patients';
import { loadCSV } from '../csv-loader';

const mockedPatient: PatientRawData[] = [
  {
    patient_id: '60',
    coordinates: '(-7.165522128346392, 6.777079232384578)',
    cluster_id: '13',
  },
  {
    patient_id: '61',
    coordinates: '(-9.385521730688493, 5.825364355168432)',
    cluster_id: '8',
  },
  {
    patient_id: '62',
    coordinates: '(-10.339938290911057, 3.342519830062648)',
    cluster_id: '17',
  },
  {
    patient_id: '63',
    coordinates: '(-5.911541191020675, 8.786440172113451)',
    cluster_id: '13',
  },
];

const mocks = vi.hoisted(() => {
  return {
    loadCSV: vi.fn(),
  };
});

vi.mock('../csv-loader', () => {
  return {
    loadCSV: mocks.loadCSV,
  };
});

vi.mocked(loadCSV).mockResolvedValue(mockedPatient);

describe('get-patient', async () => {
  const results = await getPatients();

  it('should return correct format', async () => {
    expect(results[0]).toEqual({
      patientId: 60,
      clusterId: 13,
      coordinates: {
        x: -7.165522128346392,
        y: 6.777079232384578,
      },
    } satisfies Patient);
  });

  it('should match snapshot', async () => {
    expect(results).toMatchSnapshot();
  });
});
