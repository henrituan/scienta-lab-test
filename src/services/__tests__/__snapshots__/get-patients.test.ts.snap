// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`get-patient > should match snapshot 1`] = `
[
  {
    "clusterId": 13,
    "coordinates": {
      "x": -7.165522128346392,
      "y": 6.777079232384578,
    },
    "patientId": 60,
  },
  {
    "clusterId": 8,
    "coordinates": {
      "x": -9.385521730688493,
      "y": 5.825364355168432,
    },
    "patientId": 61,
  },
  {
    "clusterId": 17,
    "coordinates": {
      "x": -10.339938290911057,
      "y": 3.342519830062648,
    },
    "patientId": 62,
  },
  {
    "clusterId": 13,
    "coordinates": {
      "x": -5.911541191020675,
      "y": 8.786440172113451,
    },
    "patientId": 63,
  },
]
`;
